Basic-Car-Maintenance/Shared/BasicCarMaintenanceApp.swift

```diff
@@ -7,16 +7,27 @@
 
 import FirebaseCore
 import SwiftUI
+import SwiftData
 
 @main
 struct BasicCarMaintenanceApp: App {
     @State private var actionService = ActionService.shared
     @UIApplicationDelegateAdaptor(AppDelegate.self) var delegate
+    let container: ModelContainer
+    
+    init() {
+        do {
+            self.container = try ModelContainer(for: AcknowledgedAlert.self)
+        } catch {
+            fatalError("Failed to create ModelContainer for AcknowledgedAlert.")
+        }
+    }
     
     var body: some Scene {
         WindowGroup {
-            MainTabView()
+            MainTabView(modelContext: container.mainContext)
                 .environment(actionService)
+                .modelContainer(container)
```

Comment:
Is there a reason we're passing it via the environment, and through the initializer?
Also do we not need to do something like this anymore, because it's `SwiftData`?

```swift
@Environment(ModelContainer.self) var container
```

Reply:
1.  Passing to the view hierarchy is related to automatically update the views if something is changed
2. Passing through the initializer is because if we declared `@Environment(\.modelContext) var modelContext`
in `MainTabView` and tried to pass it through the view's initializer we will have a compiler error as `self` isn't initialized yet

> [!NOTE]
> Since I'm still new to `SwiftData`, I followed the approach in this [tutorial](https://www.hackingwithswift.com/quick-start/swiftdata/why-are-swiftdata-models-created-as-classes)

Reply:
I think you didn't link the right tutorial? the one you linked is why you use classes for the `@Model` macro

I think it only needs to be passed once.

I'm also not an expert, because none of us are, because it's so new, but I'd follow this [Apple tutorial](https://developer.apple.com/documentation/swiftdata/adding-and-editing-persistent-data-in-your-app) which passes it through the environment and doesn't actually store it in the "view model" but instead passes the context into the vm to use as a parameter

Reply:
Sorry my bad, this is the [tutorial](https://www.hackingwithswift.com/quick-start/swiftdata/how-to-use-mvvm-to-separate-swiftdata-from-your-views) I meant
I read the tutorial you mentioned and to make sure that I understand you correctly, please let me know if the following steps are what you mean:
1. you want to have `@Environment(\.modelContext) var modelContext` in `MainTabView` and not injected through the `init`
2. in the ViewModel, you want to expose both fetch and save methods to the view and whenever we need to use them we'll pass the modelContext to them as a parameter and the VM will know nothing about modelContext?

Reply:
@mikaelacaron Any thoughts?

Reply:
I'll look at this tonight

Reply:
Okay, I've posted about this on Mastodon and Twitter, and this is a divided issue right now

Can you for now, follow the [Apple tutorials](https://developer.apple.com/documentation/swiftdata) style, even though it's not specifically "MVVM"

I think most of Apple's style is making a function in the SwiftUI view, which is fine for now

Thanks for you patience!